### PR TITLE
Type chain refactor

### PR DESCRIPTION
**Pull request by markwpearce**
- Adds options to all `AstNode.getType()` calls
- Each AstNode is Responsible for adding to the the passed-in type chain
- Added Utility function to process a type chain

- [ ] Add typechain pushing to indexedGet

### CODE REVIEW

File: src/types/BscType.ts [hunk 1]
```diff
@@ -45,3 +45,12 @@ export abstract class BscType {
         throw new Error('Method not implemented.');
     }
 }
+
+
+export class TypeResolution {
```
Comment (markwpearce) [inline src/types/BscType.ts:50]: Change name to TypeChainEntry

File: src/types/BscType.ts [hunk 2]
```diff
@@ -45,3 +45,12 @@ export abstract class BscType {
...
+    constructor(public name: string, public type: BscType, public range: Range) {
+    }
+    get resolved() {
```
Comment (markwpearce) [inline src/types/BscType.ts:53]: name change: isResolved

File: src/util.ts [hunk 3]
```diff
@@ -1472,6 +1473,35 @@ export class Util {
             }]);
         }
     }
+
+    public processTypeChain(typeChain: TypeResolution[]): { cannotFindName: string; fullErrorName: string; fullChainName: string; range: Range } {
```
Comment (markwpearce) [inline src/util.ts:1477]:
```
{
 cannotFindName => missingItemName;
fullErrorName: string; //remove 
add => lastResolvedTypeName;

 fullChainName: string;
 range: Range
} 
```

File: src/interfaces.ts [hunk 4]
```diff
@@ -411,6 +411,22 @@ export interface FileLink<T> {
...
+
+export class TypeChainEntry {
+    constructor(public name: string, public type: BscType, public range: Range) {
+    }
+    get isResolved() {
+        return this.type?.isResolvable();
+    }
+}
+
+export interface TypeChainProcessResults {
```
Comment (TwitchBronBron) [inline src/interfaces.ts:426]:
```suggestion
export interface TypeChainProcessResult {
```
Results makes me think there are multiple "results", when really this represents a single "result" just with multiple properties.

**Pull request merged**